### DISCUSSION TITLE
How to only make level 3 Slides vertical.

### DISCUSSION THREAD

**Discussion by seapat** [quarto-dev/quarto-cli#5853] [Q&A] [answered]
### Description

Hey,

I was wondering if there is some workaround to let `#` and `##` both denote horizontal slides and use only `###` for vertical slides. The `slide-level` option does put `##` and `###` on the same slides if set to 2. Here is an example:

```qmd

# (horizontal) Section heading

## Horizontal slide A

### Vertical slide A1

### Vertical slide A2

## Horizontal slide B

```

**Comment by mcanouil** [comment]
No, that's not how Pandoc thus Quarto works. 🤷‍♂️

**Reply by cscheid** [reply]
If I'm reading [our docs](https://quarto.org/docs/presentations/revealjs/advanced.html#vertical-slides) correctly, this should be possible with a Lua filter that rewrites the level of headings appropriately.

**Reply by mcanouil** [reply]
Yes, changing the default slide orientation (I forgot it was possible 😅).

Although slides are defined by level 2.
Level 3 won't make new slides.

**Reply by mcanouil** [reply]
I doubled check, the Lua would need to replace what Quarto/Pandoc do when replacing headers by nested sections.

See <https://revealjs.com/vertical-slides/> and <https://pandoc.org/MANUAL.html#structuring-the-slide-show>.

> Note: in reveal.js slide shows, if slide level is 2, a two-dimensional layout will be produced, with level-1 headings building horizontally and level-2 headings building vertically. It is not recommended that you use deeper nesting of section levels with reveal.js unless you set [--slide-level=0](https://pandoc.org/MANUAL.html#option--slide-level) (which lets reveal.js produce a one-dimensional layout and only interprets horizontal rules as slide boundaries).

**Reply by cscheid** [reply]
I must be missing something. The idea would be to use a Lua filter to find the `Header 3` elements that the user wrote, and then emit something else that is properly interpreted by Pandoc. This is a general technique in writing filters: if the output is possible to generate with a different input, then you write a Lua filter to convert the "wrong" input from the user into the input that is expected by Pandoc.

So the idea isn't to change how Pandoc works. It's to convert what the user wrote into what Pandoc is expecting.

**Reply by mcanouil** [reply]
I might be also missing something ...

Let's write it down^^

- change the default direction to `vertical`
  - h1 behave as horizontal section -> `section` not nested.
  - h2 behave as vertical section -> `section` nested.
  - h3 not a slide separator (with `slide-level: 2`) -> added in the parent `section`
  
Now, the Lua filter would need to:
1. turn `h2` into `h1`
2. turn `h3` into `h2`

But I am not convinced it would work.

**Reply by cscheid** [reply, marked as answer]
I just tested this, it works:


````
---
title: "Presentation"
format:
  revealjs:
    navigation-mode: vertical
filters:
  - relevel_headings.lua
---

# (horizontal) Section heading

## Horizontal slide A

### Vertical slide A1

### Vertical slide A2

## Horizontal slide B
````

relevel_headings.lua:

```lua
function Header(heading)
  if heading.level == 3 then
    heading.level = 2
    return heading
  elseif heading.level == 2 then
    heading.level = 1
    return heading
  end
  return heading
end
```

**Reply by cscheid** [reply]
Ok, I think your point is that in this way, the slides are actually "sections", and that's true. But the horizontal/vertical navigation works as expected.

**Reply by mcanouil** [reply]
Yep, I also just tried the filter^^

It's really not an intuitive use of Reveal.js structure (independently of how Pandoc uses markdown to creates the HTML sections).

**Reply by cderv** [reply]
What I am not sure to understand here is that 

* You are transforming `###` to `##` 
* You are transforming `##` to `#`
* You are leaving `#` untouched

So basically, Using this Lua filter is equivalent to write directly 

````markdown
# (horizontal) Section heading

# Horizontal slide A

## Vertical slide A1

## Vertical slide A2

# Horizontal slide B
```

Is this really what we want here ? This is useful for a presentation where wouldn't want to change level in `.qmd` source (because in HTML , the tags and styling will not be the level from the markdown as we change it). 

I would say, no Lua filter is needed to create vertical slides, just only using two levels of headers in document is simpler !

Though `slide-level` can help, but with different results...  

`revealjs` format with Pandoc works on two level only. Especially for vertical slides: 
* One level for Sections slides, which will be horizontal
* Other levels (above `slide-level`) will be used as vertical

This means that 
```markdown
---
title: "Presentation"
format:
  revealjs:
    navigation-mode: vertical
    slide-level: 3
---

# (horizontal) Section heading

## Horizontal slide A

### Vertical slide A1

### Vertical slide A2

## Horizontal slide B
```

Would work for `###` to create some slides. 

However, only one level can be used as horizontal slide by Pandoc. 

* `#` will do horizontal slide
* `##` will do vertical 
* `###` will do vertical too

This is why the note they shared as stated in above comment https://github.com/quarto-dev/quarto-cli/discussions/5853#discussioncomment-6126290

Using `navigation-mode: linear` (our default) will make a 1D layout of slides which `#`, `##` and `###` all creating slides. 

So overall, when using vertical slides, solutions is to use only two levels I guess!

**Reply by mcanouil** [reply]
@cderv The use of level 3 puzzled me as well 😅
The only useful thing that the various level brings is an IDE feature, _i.e._, you have more fine grained control on parts you want to collapse.

PS: there is some code blocks delimitation issues in your post.

**Reply by seapat** [reply]
Well, my idea was to make a "traditional" presentation that goes from left to right. With slides separated into sections. The level 3 slides are supposed to be used for optional slides, i.e. to answer anticipated questions, they should be the only ones that go vertical. I hope that makes it more clear, why asked for that. 

In hindsight, you probably could just style sections and normal slides differently while using the same heading. Sorry, but I haven't used quarto nor reveal before ...

**Reply by cderv** [reply]
> Well, my idea was to make a "traditional" presentation that goes from left to right. With slides separated into sections. The level 3 slides are supposed to be used for optional slides, i.e. to answer anticipated questions, they should be the only ones that go vertical. I hope that makes it more clear, why asked for that.

Unfortunaly I don't think Pando allows that for revealjs. The horizontal slide will always be a section slide with class `title-slide`. You would need to post process HTML to remove this class (or add Revealjs hook to tweak it). 

Pandoc will always to when vertical 

* As it says in its note, Pandoc works best with 2 nested levels really.
* `slide-level` what is the level to use for regular slides  
* lower level will be used as section slides. E.g `slide-level: 2` means  `##` slide, and `#` section slide. 
* Vertical modes happens with the lowest level. e.g. `slide-level: 3` means `###` creates regular slides, `##` creates section slides (as explained in bullet point just above), `#` creates section two, and will be used as the horizontal slide. 

I don't think there is currently any other as this is how the Pandoc HTML writer works for revealjs. 🤔

**Reply by mcanouil** [reply]
The only solution I see would be to use raw HTML block.

````qmd
```{=html}
<section>Horizontal Slide</section>
<section>
  <section>Vertical Slide 1</section>
  <section>Vertical Slide 2</section>
</section>
```
````

Unfortunately, Pandoc nest the bloc into section which messes up the whole thing.

```
<section class="slide level6">

<section>Horizontal Slide</section>
<section>
  <section>Vertical Slide 1</section>
  <section>Vertical Slide 2</section>
</section>
</section>
```

So an idea could be to write a Lua filter that removes/flattens `<section class="slide level6"> </section>` (I tried a bit yesterday, but unsuccessful so far to get the intended results).

**Reply by cderv** [reply]
> The only solution I see would be to use raw HTML block.

This would mean not using Pandoc or Quarto to create the revealjs presentation.  Or if to do that, it would be through a custom writer for revealjs. Pandoc's currently does it as a specific handling of the internal HTML writer
